File: tutorial/go2_tutorial/src/go2_cruising_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "go2_tutorial_inter/srv/cruising.hpp"

using namespace std::chrono_literals;

// 定義一個 Go2CruisingClient 類別，繼承自 rclcpp::Node
class Go2CruisingClient : public rclcpp::Node{
public:
    // 構造函數，初始化節點名稱並創建服務客戶端
    Go2CruisingClient(): Node("go2_cruising_client"){
        RCLCPP_INFO(this->get_logger(), "Go2CruisingClient created!");
        cru_client_ = this->create_client<go2_tutorial_inter::srv::Cruising>("cruising");
    }

    // 連接服務器的方法，等待服務可用
    bool connect_server(){
      while (!cru_client_->wait_for_service(1s))
      {
        if (!rclcpp::ok()){
          return false; // 如果 ROS2 系統已關閉，返回 false
        }
        RCLCPP_INFO(this->get_logger(), "service connecting...");
      }
      return true; // 服務可用時返回 true
    }

    // 發送請求的方法，傳入一個整數參數 flag
    rclcpp::Client<go2_tutorial_inter::srv::Cruising>::FutureAndRequestId send_request(int32_t flag){
      auto req_ = std::make_shared<go2_tutorial_inter::srv::Cruising_Request>();
      req_->flag = flag; // 設置請求的 flag 參數
      return cru_client_->async_send_request(req_); // 發送異步請求
    }

private:
    // 定義服務客戶端的共享指針
    rclcpp::Client<go2_tutorial_inter::srv::Cruising>::SharedPtr cru_client_;
};

// 主函數，程序入口點
int main(int argc, char * argv[])
{
  if (argc != 2){
    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "please commit an int"); // 檢查參數數量是否正確
    return 1;
  }

  rclcpp::init(argc, argv); // 初始化 ROS2 系統

  auto client_ = std::make_shared<Go2CruisingClient>(); // 創建 Go2CruisingClient 節點
  auto flag = client_->connect_server(); // 嘗試連接服務器

  if (!flag){
    return 1; // 如果連接失敗，退出程序
  }

  auto response_future = client_->send_request(atoi(argv[1])); // 發送請求，參數來自命令行

  // 等待請求結果
  if(rclcpp::spin_until_future_complete(client_, response_future) == rclcpp::FutureReturnCode::SUCCESS){
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "SUCCESS"); // 請求成功
    auto response_ = response_future.get(); // 獲取響應
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "robot at:(%.2f, %.2f)", response_->point.x, response_->point.y); // 輸出響應結果
  }
  else{
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "FAIL"); // 請求失敗
  }

  rclcpp::shutdown(); // 關閉 ROS2 系統
  return 0;
}

File: tutorial/go2_tutorial_py/go2_tutorial_py/go2_cruising_client.py
# 匯入 ROS2 相關模組
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
import sys
from go2_tutorial_inter.srv import Cruising
from rclpy.task import Future

# Go2 巡航客戶端節點類別
class Go2CruisingClient(Node):
    # 初始化函式，建立服務客戶端
    def __init__(self):
        super().__init__("go2_cruising_client")
        self.client = self.create_client(Cruising, "cruising")  # 建立服務客戶端

    # 連線到服務端，若失敗則重試
    def connect_server(self):
        while not self.client.wait_for_service(1.0):
            if not rclpy.ok():
                return False
            get_logger("rclpy").error("service not available, waiting again...")
        self.get_logger().info("service available")
        return True

    # 發送請求到服務端
    def send_request(self, flag) -> Future:
        req = Cruising.Request()
        req.flag = int(flag)  # 設定請求的 flag
        return self.client.call_async(req)  # 非同步呼叫服務

# 主程式進入點
def main():
    # 檢查參數數量
    if len(sys.argv) != 2:
        get_logger("rclpy").error("please input 1 int argument")
        return

    rclpy.init()

    cru_client = Go2CruisingClient()  # 建立巡航客戶端節點

    flag = cru_client.connect_server()  # 嘗試連線到服務端
    if not flag:
        get_logger("rclpy").error("connect server failed")
        return

    future = cru_client.send_request(int(sys.argv[1]))  # 發送請求

    rclpy.spin_until_future_complete(cru_client, future)  # 等待請求完成
    if future.done():
        response : Cruising.Response = future.result()
        get_logger("rclpy").info(f"current position: {response.point}")  # 輸出當前位置
    else:
        get_logger("rclpy").error("service call failed")
    
    rclpy.shutdown()

if __name__ == "__main__":
    main()